src/map/entities/mobentity.cpp

```diff
@@ -924,6 +913,58 @@ void CMobEntity::DropItems(CCharEntity* PChar)
                     return;
             }
         }
+        // Todo: Avatarite and Geode drops during day/weather. Much higher chance during weather than day.
+        // Item element matches day/weather element, not mob crystal. Lv80+ xp mobs can drop Avatarite.
+        // Wiki's have conflicting info on mob lv required for Geodes. One says 50 the other 75. I think 50 is correct.
+
+        uint8 effect = 0; // Begin Adding Crystals
+        
+        if (conquest::GetRegionOwner(PChar->loc.zone->GetRegionID()) <= 2 && m_Element > 0) // Valid Signet Zone
+        {
+            effect = 1;
+        }
+        else if (PChar->loc.zone->GetRegionID() >= 28 && PChar->loc.zone->GetRegionID() <= 32 && m_Element > 0) // Valid Sanction Zone
+        {
+            effect = 2;
+        }
+        else if (PChar->loc.zone->GetRegionID() >= 33 && PChar->loc.zone->GetRegionID() <= 40 && m_Element > 0) // Valid Sigil Zone
+        {
+            effect = 3;
+        }
+        int crystalRolls = 0;
+        PChar->ForParty([this, &crystalRolls, &effect](CBattleEntity* PMember)
+        {
+            switch(effect)
+            {
+                case 1:
+                    if (PMember->StatusEffectContainer->HasStatusEffect(EFFECT_SIGNET) && PMember->getZone() == getZone() && distance(PMember->loc.p, loc.p) < 100)
+                    {
+                        crystalRolls++;
+                    }
+                    break;
+                case 2:
+                    if (PMember->StatusEffectContainer->HasStatusEffect(EFFECT_SANCTION) && PMember->getZone() == getZone() && distance(PMember->loc.p, loc.p) < 100)
+                    {
+                        crystalRolls++;
+                    }
+                    break;
+                case 3:
+                    if (PMember->StatusEffectContainer->HasStatusEffect(EFFECT_SIGIL) && PMember->getZone() == getZone() && distance(PMember->loc.p, loc.p) < 100)
+                    {
+                        crystalRolls++;
+                    }
+                    break;
+                default:
+                    break;
+            }
+        });
```

Comment:
Let's simplify this block as each case is checking:

- if the member is in zone
- if the member is within distance

```cpp

PChar->ForParty([this, &crystalRolls, &effect](CBattleEntity* PMember)
{
    If (PMember->getZone() == getZone() && distance(PMember->loc.p, loc.p) < 100)
    {
        switch (effect)
        {
            case 1:
                if (PMember->StatusEffectContainer->HasStatusEffect(EFFECT_SIGNET)
                {
                    crystalRolls++;
                }
                break;
            // other cases similar..
        }
    }
});
```

Reply:
Each case except the default case, and then for each member we don't even bother checking zone and distance unless they also have the status.